### ISSUE TITLE
Problem with shared library compiled without cuda support

### ISSUE THREAD

**Issue by hellbago** [FloopCZ/tensorflow_cc#169]
I'm using a freezed graph in c++ project with tensorflow c++ api compiled using this repo. I have no issue when the library is compiled as shared and with cuda support enabled. Since I need to use the software in machines without a gpu, I've also compiled tensorlfow as shared library but without cuda support. With this configuration I obtain a runtime error when the network runs

```
2019-04-12 17:29:37.169345: E tensorflow/core/framework/tensor.cc:122] Input size was 44 and expected 4

2019-04-12 17:29:37.177419: E tensorflow/core/framework/tensor.cc:122] Input size was 44 and expected 1024

2019-04-12 17:29:37.177719: E tensorflow/core/framework/tensor.cc:122] Input size was 44 and expected 4

2019-04-12 17:29:37.177767: E tensorflow/core/framework/tensor.cc:122] Input size was 44 and expected 4

2019-04-12 17:29:37.177801: E tensorflow/core/framework/tensor.cc:122] Input size was 44 and expected 4

2019-04-12 17:29:37.177953: E tensorflow/core/framework/tensor.cc:122] Input size was 44 and expected 1152

2019-04-12 17:29:37.178237: E tensorflow/core/framework/tensor.cc:122] Input size was 44 and expected 1024

2019-04-12 17:29:37.179021: E tensorflow/core/framework/tensor.cc:122] Input size was 44 and expected 4

2019-04-12 17:29:37.179071: E tensorflow/core/framework/tensor.cc:122] Input size was 44 and expected 4

2019-04-12 17:29:37.180486: E tensorflow/core/framework/tensor.cc:122] Input size was 44 and expected 192

2019-04-12 17:29:37.180966: E tensorflow/core/framework/tensor.cc:122] Input size was 44 and expected 384

2019-04-12 17:29:37.181072: E tensorflow/core/framework/tensor.cc:122] Input size was 44 and expected 4

2019-04-12 17:29:37.181402: E tensorflow/core/framework/tensor.cc:122] Input size was 44 and expected 320

2019-04-12 17:29:37.181584: E tensorflow/core/framework/tensor.cc:122] Input size was 44 and expected 768

2019-04-12 17:29:37.181710: E tensorflow/core/framework/tensor.cc:122] Input size was 44 and expected 128

2019-04-12 17:29:37.181985: E tensorflow/core/framework/tensor.cc:122] Input size was 44 and expected 256

2019-04-12 17:29:37.182039: E tensorflow/core/framework/tensor.cc:122] Input size was 44 and expected 4

2019-04-12 17:29:37.182243: E tensorflow/core/framework/tensor.cc:122] Input size was 44 and expected 4

2019-04-12 17:29:37.182310: E tensorflow/core/framework/tensor.cc:122] Input size was 44 and expected 4

[libprotobuf FATAL external/protobuf_archive/src/google/protobuf/map.h:1058] CHECK failed: it != end(): key not found: /job:localhost/replica:0/task:0/device:CPU:0

terminate called after throwing an instance of 'google::protobuf::FatalException'

  what():  CHECK failed: it != end(): key not found: /job:localhost/replica:0/task:0/device:CPU:0
```

The error seams to be related to the fact that tensorflow is unable to intercept a device (in this case the cpu) for running the graph.
Any advice on this issue or some previous experience? @FloopCZ

**Comment by FloopCZ** [comment]
Hi, you can retrieve the version from cmake. Try:
```
cat /usr/local/lib/cmake/TensorflowCC/TensorflowCCConfigVersion.cmake | grep ^set\(PACKAGE_VERSION
```

**Comment by FloopCZ** [comment]
Hi, sorry, I have never hit this issue. Anyone?
Does it happen also in the official docker images with tag `ubuntu-shared`?

**Comment by hellbago** [comment]
Output for `cat /usr/local/lib/cmake/TensorflowCC/TensorflowCCConfigVersion.cmake | grep ^set\(PACKAGE_VERSION`   is:
`set(PACKAGE_VERSION "1.12.0")`
@FloopCZ

**Comment by hellbago** [comment]
Using the latest version of the repo the problem still appears. Can you try to replicate the issue with some pretrained network? @FloopCZ

**Comment by FloopCZ** [comment]
And does it happen with the ubuntu-shared docker image as well?

**Comment by hellbago** [comment]
Sorry @FloopCZ I'm not an expert of Docker Images. My project has two main dependencies: tensorflow and opencv. I suppose using the docker image I can solve the tensorflow dependency but not the opencv opencv. Or maybe I'm wrong. At the moment i have a Cmake file that links tensorflow and opencv installed locally. Can you evetually give me some tips in order to get my project work with the docker image you provide?  Thanks

**Comment by hellbago** [comment]
> And does it happen with the ubuntu-shared docker image as well?

Yes it happens also with the ubuntu-shared docker image

**Comment by FloopCZ** [comment]
Hi, I have just tried to run the [official label_image example](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/examples/label_image) in the ubuntu-shared docker image using their [pre-trained inception network](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/examples/label_image#to-buildinstallrun) and it seems to work just fine. Can you reproduce that?
```
root@b9fe7b9a2810:/tensorflow_cc/example/build# ./example --image=data/grace_hopper.jpg --graph=data/inception_v3_20
16_08_28_frozen.pb --labels=data/imagenet_slim_labels.txt
2019-04-30 21:14:58.498002: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2019-04-30 21:14:58.547311: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 3401105000 Hz
2019-04-30 21:14:58.548688: I tensorflow/compiler/xla/service/service.cc:150] XLA service 0x5578c3d8b1a0 executing computations on platform Host. Devices:
2019-04-30 21:14:58.548760: I tensorflow/compiler/xla/service/service.cc:158]   StreamExecutor device (0): <undefined>, <undefined>
2019-04-30 21:15:05.237127: I /tensorflow_cc/example/example.cpp:251] military uniform (653): 0.834307
2019-04-30 21:15:05.237302: I /tensorflow_cc/example/example.cpp:251] mortarboard (668): 0.0218695
2019-04-30 21:15:05.237372: I /tensorflow_cc/example/example.cpp:251] academic gown (401): 0.0103581
2019-04-30 21:15:05.237417: I /tensorflow_cc/example/example.cpp:251] pickelhaube (716): 0.00800814
2019-04-30 21:15:05.237448: I /tensorflow_cc/example/example.cpp:251] bulletproof vest (466): 0.00535084
```

**Comment by hellbago** [comment]
> Hi, I have just tried to run the [official label_image example](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/examples/label_image) in the `ubuntu-shared` docker image using their [pre-trained inception network](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/examples/label_image#to-buildinstallrun) and it seems to work just fine. Can you reproduce that?
> 
> ```
> root@b9fe7b9a2810:/tensorflow_cc/example/build# ./example --image=data/grace_hopper.jpg --graph=data/inception_v3_20
> 16_08_28_frozen.pb --labels=data/imagenet_slim_labels.txt
> 2019-04-30 21:14:58.498002: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
> 2019-04-30 21:14:58.547311: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 3401105000 Hz
> 2019-04-30 21:14:58.548688: I tensorflow/compiler/xla/service/service.cc:150] XLA service 0x5578c3d8b1a0 executing computations on platform Host. Devices:
> 2019-04-30 21:14:58.548760: I tensorflow/compiler/xla/service/service.cc:158]   StreamExecutor device (0): <undefined>, <undefined>
> 2019-04-30 21:15:05.237127: I /tensorflow_cc/example/example.cpp:251] military uniform (653): 0.834307
> 2019-04-30 21:15:05.237302: I /tensorflow_cc/example/example.cpp:251] mortarboard (668): 0.0218695
> 2019-04-30 21:15:05.237372: I /tensorflow_cc/example/example.cpp:251] academic gown (401): 0.0103581
> 2019-04-30 21:15:05.237417: I /tensorflow_cc/example/example.cpp:251] pickelhaube (716): 0.00800814
> 2019-04-30 21:15:05.237448: I /tensorflow_cc/example/example.cpp:251] bulletproof vest (466): 0.00535084
> ```

I've tried this example inside the ubuntu-shared docker image and in my local machine and it works well in both cases. So the issue must be related to my .pb file. I've created my freezed file using tensorflow-gpu 1.4. As I mentioned in a previous post the network runs fine with tensorflow_cc compiled as shared lib and cuda support. Maybe the problem can be linked with some gpu stuff inside the freezed file. Do you have some suiggestions about this? @FloopCZ

**Comment by FloopCZ** [comment]
Try to use the `freeze_graph.py` from the latest `master`:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/freeze_graph.py
It should properly clear the device info from the frozen model.

**Comment by FloopCZ** [comment]
Closing for inactivity, feel free to reopen if the issue persists.

**Comment by HOYS** [comment]
Hello!

I've come across a similar issue and this is the closest I've found to someone having a similar issue.

I get errors very similar to the OP when I attempt to set the default device to the GPU.

`tensorflow/core/framework/tensor.cc:122] Input size was 44 and expected 4`

The only difference is I attempted to compile with CUDA support but have been unable to activate the GPU, only CPU.

For what its worth I have python based Tensorflow running well on my system with CUDA 10.0, RTX2060.
